### PR TITLE
Update the API of get_fcp_template_details and get_fcp_templates to return the CPC/LPAR/PCHID info

### PR DESCRIPTION
**Pull request by renpei123**
Update the API of get_fcp_template_details and get_fcp_templates to return the CPC/LPAR/PCHID info

### CODE REVIEW

File: zvmsdk/database.py [hunk 1]
```diff
@@ -1566,7 +1588,7 @@ def get_fcp_templates_details(self, template_id_list=None):
         devices_cmd = (
             "SELECT tf.fcp_id, tf.tmpl_id, tf.path, fcp.assigner_id, "
             "fcp.connections, fcp.reserved, fcp.wwpn_npiv, fcp.wwpn_phy, "
-            "fcp.chpid, fcp.state, fcp.owner, fcp.tmpl_id "
+            "fcp.chpid, fcp.pchid, fcp.state, fcp.owner, fcp.tmpl_id "
```
Comment (dongyanyang) [inline zvmsdk/database.py:1591]: comment in line 1570 also needs to be updated.

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -1323,6 +1324,20 @@ def _update_template_fcp_statistics_usage(self, statistics_usage,
                         template_id][path_id]["CHPIDs"].update({chpid: []})
                 statistics_usage[
                     template_id][path_id]["CHPIDs"][chpid].append(fcp_id)
+            if pchid:
+                if not statistics_usage[template_id][path_id]["PCHID-CHPID"].get(pchid, None):
+                    statistics_usage[
+                        template_id][path_id]["PCHID-CHPID"].update({pchid: []})
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1330]: why using a list as the value? pchid:[], shouldn't it be a string? 
chpid and pchid is a 1v1 relationship, no need to use []

File: zvmsdk/volumeop.py [hunk 3]
```diff
@@ -1323,6 +1324,20 @@ def _update_template_fcp_statistics_usage(self, statistics_usage,
...
+                    statistics_usage[
+                        template_id][path_id]["total_count"].update({pchid: []})
+                    statistics_usage[
+                        template_id][path_id]["available_count"].update({pchid: []})
+                statistics_usage[
+                    template_id][path_id]["PCHID-CHPID"][pchid].append(chpid)
+                statistics_usage[
+                    template_id][path_id]["total_count"][pchid].append(fcp_id)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1338]: here you use the total_count actually to store the fcp_id list on this pchid.
Although in later stage, you changed this value to the count, it is still confusing when reading this part of code.

suggest to use other name to store this info, you can later pop it out if not useful any more.
@jackydalong can also give some advice.

File: zvmsdk/volumeop.py [hunk 4]
```diff
@@ -1323,6 +1324,20 @@ def _update_template_fcp_statistics_usage(self, statistics_usage,
...
+                statistics_usage[
+                    template_id][path_id]["available_count"][pchid].append(fcp_id)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1340]: same as above

**Review by dongyanyang — CHANGES_REQUESTED** [review]

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/volumeop.py (hunk 4 above)
Comment (renpei123) [reply zvmsdk/volumeop.py]: Done in https://github.com/openmainframeproject/feilong/pull/732#discussion_r1281718751

File: zvmsdk/volumeop.py [hunk 5]
```diff
@@ -1375,6 +1389,10 @@ def _shrink_fcp_list_in_statistics_usage(self, statistics_usage):
                     fcp_list = fcps
                     template_statistics[path]['CHPIDs'][chpid] = (
                         utils.shrink_fcp_list(fcp_list))
+                # shrink for each PCHID-CHPID
+                for pchid, chpid in template_statistics[
```
[the remark this answers is not in the document]
Comment (renpei123) [reply zvmsdk/volumeop.py]: Removed.

File: zvmsdk/volumeop.py [hunk 6]
```diff
@@ -1323,6 +1324,32 @@ def _update_template_fcp_statistics_usage(self, statistics_usage,
...
+                    statistics_usage[
+                        template_id][path_id]["available_count"].update({pchid: 0})
+                statistics_usage[
+                    template_id][path_id]["PCHID-CHPID"][pchid] = chpid
+                statistics_usage[
+                    template_id][path_id]["total_count"][pchid] += 1
+                if fcp_id in statistics_usage[template_id][path_id]["available"]:
+                    statistics_usage[
+                        template_id][path_id]["available_count"][pchid] += 1
+            else:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1342]: no need to keep this else.
Also defect in line 1346 and line 1349

File: zvmsdk/volumeop.py [hunk 7]
```diff
@@ -1501,6 +1539,18 @@ def get_fcp_templates(self, template_id_list=None, assigner_id=None,
...
+        # Get zhyp_info
+        zhypinfo = zvmutils.get_zhypinfo()
+        for item in ret:
+            # Add cpc_sn info
+            item["cpc_sn"] = zvmutils.get_cpc_sn(zhypinfo=zhypinfo)
+            # Add cpc_name info
+            item["cpc_name"] = zvmutils.get_cpc_name(zhypinfo=zhypinfo)
+            # Add lpar info
+            item["lpar"] = zvmutils.get_lpar_name(zhypinfo=zhypinfo)
+            # Add PCHIDs info
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1552]: cpc_sn, name, lpar, all these 3 items are same for all templates, no matter to get once for each template.

File: zvmsdk/volumeop.py [hunk 8]
```diff
@@ -1730,10 +1790,27 @@ def get_fcp_templates_details(self, template_id_list=None, raw=False,
...
+                        if path_detail['PCHID-CHPID']:
+                            pchids.append(list(path_detail['PCHID-CHPID'].keys())[0])
+                    base_info.update({"PCHIDs": pchids})
+                    zhypinfo = zvmutils.get_zhypinfo()
+                    # Add cpc_sn info
+                    base_info.update({"cpc_sn": zvmutils.get_cpc_sn(zhypinfo=zhypinfo)})
+                    # Add cpc_name info
+                    base_info.update({"cpc_name": zvmutils.get_cpc_name(zhypinfo=zhypinfo)})
+                    # Add lpar info
+                    base_info.update({"lpar": zvmutils.get_lpar_name(zhypinfo=zhypinfo)})
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1805]: same as above, no need to get these info once for each template

File: zvmsdk/volumeop.py [hunk 9]
```diff
@@ -1730,10 +1790,27 @@ def get_fcp_templates_details(self, template_id_list=None, raw=False,
...
                 else:
                     # some templates do not have fcp devices or do not have
                     # valid fcp in zvm, so do not have statistics_usage data
                     base_info.update({"statistics": {}})
+                    base_info.update({"PCHIDs": []})
+                    base_info.update({"cpc_sn": ''})
+                    base_info.update({"cpc_name": ''})
+                    base_info.update({"lpar": ''})
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1813]: although no FCP device, the cpc_sn,name,lpar still should be returned with correct value instead of empty string.

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/tests/unit/test_volumeop.py [hunk 10]
```diff
@@ -858,9 +858,19 @@ def test_sync_fcp_table_with_zvm(self, get_pchid_by_chpid):
         finally:
             self.db_op.bulk_delete_from_fcp_table(fcp_id_list)
 
+    @mock.patch('zvmsdk.utils.get_zhypinfo')
     @mock.patch.object(uuid, 'uuid1')
-    def test_create_fcp_template(self, get_uuid):
+    def test_create_fcp_template(self, get_uuid, mock_get_zhypinfo):
```
Comment (dongyanyang) [inline zvmsdk/tests/unit/test_volumeop.py:863]: this PR only change the get_xxx function, why would you change UT of create_fcp_template?

Comment (renpei123) [reply zvmsdk/tests/unit/test_volumeop.py]: In `test_create_fcp_template`, there's a step to call `get_fcp_templates`. 
Made change in `test_create_fcp_template` is caused by return value for `get_fcp_templates` changed.

File: zvmsdk/volumeop.py (hunk 9 above)
Comment (renpei123) [reply zvmsdk/volumeop.py]: Done.

File: zvmsdk/volumeop.py (hunk 8 above)
Comment (renpei123) [reply zvmsdk/volumeop.py]: Done.

File: zvmsdk/volumeop.py (hunk 7 above)
Comment (renpei123) [reply zvmsdk/volumeop.py:1597]: Done.

File: zvmsdk/volumeop.py (hunk 6 above)
Comment (renpei123) [reply zvmsdk/volumeop.py]: Done.

File: zvmsdk/volumeop.py [hunk 11]
```diff
@@ -1358,6 +1359,31 @@ def _update_template_fcp_statistics_usage(self, statistics_usage,
...
+                    statistics_usage[
+                        template_id][path_id]["available_count"].update({pchid: 0})
+                statistics_usage[
+                    template_id][path_id]["PCHIDs"][pchid] = chpid
+                statistics_usage[
+                    template_id][path_id]["total_count"][pchid] += 1
+                if fcp_id in statistics_usage[template_id][path_id]["available"]:
+                    statistics_usage[
+                        template_id][path_id]["available_count"][pchid] += 1
+                if len(statistics_usage[template_id][path_id]["PCHIDs"]) == 0:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:1377]: code from line 1377 to line 1385 is not needed.

The value is already initiated in line 1255, 1259, 1267-1268

Comment (renpei123) [reply zvmsdk/volumeop.py]: Done.

**Review by dongyanyang — APPROVED** [review]

**Review by bjhuangr — APPROVED** [review]

**Pull request merged**